aip/0162.md

```diff
@@ -0,0 +1,229 @@
+  Book book = 1;
+
+  // The date that this revision was created.
+  google.protobuf.Timestamp create_time = 2;
+}
+```
+
+- Revision objects **should** include a representation of the resource as of
+  that revision.
+  - The `string name` field in the resource **must** include an explicit
+    revision ID.
+  - If providing the full resource is expensive or infeasible, the revision
+    object **may** include a `string name` field instead, which **must**
+    include the resource name and an explicit revision ID.
+- Revision objects **must** include the timestamp when the revision was
+  created, which **should** be called `create_time`.
+- Revision objects **may** include other information, such as a description or
+  other metadata, if desired.
+
+### Snapshotting revisions
+
+Depending on the resource, different APIs may have different strategies for
+when to snapshot a new revision, such as:
+
+- Snapshot a new revision any time that there is a change
+- Snapshot a new revision when something important happens
+- Snapshot a new revision when the user specifically asks
+
+APIs **may** use any of these strategies. APIs that want to snapshot a revision
+on user request **should** handle this with a `Snapshot` custom method:
+
+```proto
+rpc SnapshotBook(SnapshotBookRequest) returns (Book) {
+  option (google.api.http) = {
+    post: "/v1/{name=publishers/*/books/*}:snapshot"
+    body: "*"
+  }
+}
+```
+
+- The method **must** use the `POST` HTTP verb.
+- The method **should** return the resource, and the resource name **must**
+  include the revision ID.
+
+The request message **must** include the `name` field, and **may** include any
+additional metadata about the revision (such as a description), which
+**should** then be included in revision objects.
+
+### Rollback
+
+A common use case for a resource with a revision history is the ability to roll
+back to a given revision. APIs **should** handle this with a `Rollback` custom
+method:
+
+```proto
+rpc RollbackBook(RollbackBookRequest) returns (Book) {
+  option (google.api.http) = {
+    post: "/v1/{name=publishers/*/books/*}:rollback"
+    body: "*"
+  }
+}
+```
+
+- The method **must** use the `POST` HTTP verb.
+- The method **should** return the resource, and the resource name **must**
+  include the revision ID.
+
+```proto
+message RollbackBookRequest {
+  // The book being rolled back.
+  string name = 1;
+
+  // The revision ID to roll back to.
+  // It must be a revision of the same book.
+  //
+  //   Example: c7cfa2a8
+  string revision_id = 2;
+}
+```
+
+- The request message **must** include a `revision_id` field.
+  - The API **must** fail the request with `NOT_FOUND` if the revision does not
+    exist on that resource.
+
+**Note:** When rolling back, the API should return a _new_ revision of the
+resource with a _new_ revision ID, rather than reusing the original ID. This
+avoids problems with representing the same revision being active for multiple
+ranges of time.
+
```

Comment:
Note: Add a provision saying that `@` with no version means "latest, but return the name with the revision specifically included".

Reply:
Done.